torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,375 @@
+                                     "been registered without a tensor.")
+        return self(tensor)
+
+    def original(self):
+        # TODO Write docs
+        try:
+            return self.last_parametrization().orig
+        except AttributeError:
+            raise ValueError("The parametrization is not registered to a tensor.")
+
+    def last_parametrization(self):
+        # TODO Write docs
+        last = self
+        while last.is_chained():
+            last = last.orig_param
+        return last
+
+    def chain(self, parametrization):
+        r""" It modifies the current parametrization, chaining a new one to it.
+        We have that ``p2.evaluate(p1.evaluate(t)) == p2.chain(p1).evaluate(t)``
+
+        The original parametrization is modified after chain is applied to it
+
+        Args:
+            parametrization (nn.Parametrization): the parametrization to be chained
+        Returns:
+            Module: The original module with the parametrization module appended to it
+        """
+        if not isinstance(parametrization, Parametrization):
+            raise ValueError("The object '{}' is not a Parametrization"
+                             .format(parametrization))
+
+        if self.is_registered() and parametrization.is_registered():
+            raise ValueError("Both parametrizations are registered to a tensor. "
+                             "The chaining is ambiguous.")
+        last = self.last_parametrization()
+        is_parametrized = hasattr(last, "orig")
+        if is_parametrized:
+            # We move the orig tensor to the new last parametrization
+            orig = last.orig
+            delattr(last, "orig")
+
+        _set_parametrization(last, parametrization, "orig", set_cache=False)
+        if is_parametrized:
+            new_last = parametrization.last_parametrization()
+            if isinstance(orig, Parameter):
+                new_last.register_parameter("orig", orig)
+            else:
+                new_last.register_buffer("orig", orig)
+        return self
+
+    def is_chained(self):
+        # TODO Write docs
+        return is_parametrized(self, "orig")
+
+    def is_registered(self):
+        # TODO Write docs
+        try:
+            self.original()
+            return True
+        except ValueError:
+            return False
+
+
+def _set_parametrization(module, parametrization, tensor_name, set_cache):
+    r""" Sets up the parametrization mechanism used by parametrizations.
+    This works by substituting the class of the module by a class
+    that extends it and makes `tensor_name` into a property. It also
+    regsters the parametrization under `tensor_name + "_param"`.
+    If ``set_cache == True``, it also sets up the caching mechanism
+    """
+
+    # Define the getter and set the caching mechanism
+    if set_cache:
+        def get_parametrized(module):
+            cache = getattr(module, tensor_name + "_cache")
```

Comment:
I don't follow this caching logic. Why put it on the original module and not the `Parametrization` that has produced this? Also, if you put the caching logic into the `Parametrization` then the getter can always use the same code. Finally, it might be worth considering supporting `setattr` so that people can easily remove parametrizations by assigning a `Parameter` to a parametrized one, or overwriting the current one.

Reply:
The getter can always be the same code, I just separated it to simplify it, but I could just leave it as the same code. The caching logic is in the original module because there should be just one cache per parametrized element. Otherwise, if you put 8 parametrizations on a buffer you would end up with 8 different cached tensors.

The `setattr`... I am not sure how I feel about it. Is there any use-case where one would want to rewrite a parameter or a tensor with a completely new tensor, rather than removing first the parametrization with  `P.remove()`? It would also require monkey-patching setattr, but I guess that that'd be alright.

Reply:
Oh good point about not saving the intermediate parametrizations! But you could still work around this by doing this: when the `Sequential`-like module for parametrizations is called, disable the caching for the duration of `forward`, to prevent the inner-parametrizations from saving anything. However, once forward is over, the caching is restored and it's the `Sequential` that will never get its `forward` called, because the cache lives in this object.

Reply:
See long answer above on using a `Sequential`-like module.